virtualizarr/readers/zarr.py

```diff
@@ -0,0 +1,444 @@
+    # This currently fails for local filepaths (ie. tests):
+    # *** TypeError: Filesystem needs to support async operations.
+    # https://github.com/zarr-developers/zarr-python/issues/2554
+
+    # use UPath for combining store path + chunk key when building chunk manifests
+    store_path = UPath(filepath)
+
+    if reader_options is None:
+        reader_options = {}
+
+    zg = zarr.open_group(
+        filepath, storage_options=reader_options.get("storage_options"), mode="r"
+    )
+
+    zarr_arrays = [val for val in zg.keys()]
+
+    # mypy typing
+    if loadable_variables is None:
+        loadable_variables = set()
+
+    if drop_variables is None:
+        drop_variables = set()
+
+    missing_vars = set(loadable_variables) - set(zarr_arrays)
+    if missing_vars:
+        raise ValueError(
+            f"Some loadable variables specified are not present in this zarr store: {missing_vars}"
+        )
+
+    virtual_vars = list(
+        set(zarr_arrays) - set(loadable_variables) - set(drop_variables)
+    )
+
+    virtual_variable_mapping = {
+        f"{var}": construct_virtual_array(
+            zarr_group=zg, var_name=var, store_path=store_path
+        )
+        for var in virtual_vars
+    }
+
+    coord_names = list(
+        set(
+            item
+            for tup in [
+                virtual_variable_mapping[val].dims for val in virtual_variable_mapping
+            ]
+            for item in tup
+        )
+    )
+
+    non_loadable_variables = list(set(virtual_vars).union(set(drop_variables)))
+
+    loadable_vars, indexes = open_loadable_vars_and_indexes(
+        filepath,
+        loadable_variables=loadable_variables,
+        reader_options=reader_options,
+        drop_variables=non_loadable_variables,
+        indexes=indexes,
+        group=group,
+        decode_times=decode_times,
+    )
+
+    return construct_virtual_dataset(
+        virtual_vars=virtual_variable_mapping,
+        loadable_vars=loadable_vars,
+        indexes=indexes,
+        coord_names=coord_names,
+        attrs=zg.attrs.asdict(),
+    )
+
+
+async def get_chunk_size(zarr_group: zarr.Group, chunk_key: PosixPath) -> int:
+    # User zarr-pythons `getsize` method to get bytes per chunk
+    return await zarr_group.store.getsize(chunk_key)
+
+
+async def chunk_exists(zarr_group: zarr.Group, chunk_key: PosixPath) -> bool:
+    # calls zarr-pythons `exists` to check for a chunk
+    return await zarr_group.store.exists(chunk_key)
+
+
+async def list_store_keys(zarr_group: zarr.Group) -> list[str]:
+    # Lists all keys in a store
+    return [item async for item in zarr_group.store.list()]
+
+
+async def get_chunk_paths(
+    zarr_group: zarr.Group, array_name: str, store_path: upath.core.UPath
+) -> dict:
+    chunk_paths = {}
+
+    # Can we call list() on an array instead of the entire store?
+    store_keys = zarr.core.sync.sync(list_store_keys(zarr_group))
+
+    for item in store_keys:
```

Comment:
The async way to do this is.

```suggestion
    async for item in zarr_group.store.list():
```

> # Can we call list() on an array instead of the entire store?

You probably want `list_dir` then!